README.md
# Picture link

[![Latest Version](https://img.shields.io/github/release/divarsoy/picturelink.svg?style=flat-square)](https://github.com/divarsoy/picturelink/releases)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Build Status](https://img.shields.io/travis/divarsoy/picturelink/master.svg?style=flat-square)](https://travis-ci.org/divarsoy/picturelink)
[![Coverage Status](https://img.shields.io/scrutinizer/coverage/g/divarsoy/picturelink.svg?style=flat-square)](https://scrutinizer-ci.com/g/divarsoy/picturelink/code-structure)
[![Quality Score](https://img.shields.io/scrutinizer/g/divarsoy/picturelink.svg?style=flat-square)](https://scrutinizer-ci.com/g/divarsoy/picturelink)
[![Total Downloads](https://img.shields.io/packagist/dt/divarsoy/picturelink.svg?style=flat-square)](https://packagist.org/packages/divarsoy/picturelink)


This package aims to simplify the process of making responsive images.
## Prerequisites

* Node.js v0.10+ or io.js
* libvips v7.40.0+ (7.42.0+ recommended)
* C++11 compatible compiler such as gcc 4.6+, clang 3.0+ or MSVC 2013

## Install libvips

run the following as a user with sudo access:


``` bash
$ curl -s https://raw.githubusercontent.com/lovell/sharp/master/preinstall.sh | sudo bash -
```
or check out intall instructions at https://www.npmjs.com/package/sharp

## Install NPM packages

``` bash
$ npm install
```
## Install Picturelink

Via Composer

``` bash
$ composer require divarsoy/picturelink
```

## Usage
Setup source and destination folder in gulpfile.js
Then run

``` bash
$ gulp img
```
to resize all images in the source folder and output them to the destination folder. You can also setup a watcher with

``` bash
$ gulp watch[img]
```

You can now use picturelink in your php scripts with
``` php
$picturelink = new Divarsoy\Picturelink();
echo $picturelink->link( '/img/test.jpg', array(320,480,768,1024));
```
It will automatically download the lowest resolution image for the browser size.

## Testing

``` bash
$ phpunit
```

## Contributing

Please see [CONTRIBUTING](https://github.com/divarsoy/picturelink/blob/master/CONTRIBUTING.md) for details.

## Credits

- [Dag A. Ivarsøy](https://github.com/divarsoy)
- [All Contributors](https://github.com/divarsoy/picturelink/contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
#Changelog
All Notable changes to `Picturelink` will be documented in this file.

## NEXT - YYYY-MM-DD

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Remove
- Nothing

### Security
- Nothing


tests/PicturelinkTest.php
<?php
/**
 * Created by PhpStorm.
 * User: Dag
 * Date: 08/06/2015
 * Time: 12:39
 */

namespace Divarsoy\Picturelink\Test;
use Divarsoy\Picturelink\Picturelink;

class PicturelinkTest extends \PHPUnit_Framework_TestCase {

    /**
     * @var Picturelink
     */
    private $picturelink;

    public function setUp()
    {
        $this->picturelink = new Picturelink();
    }

    public function testMakeLinkReturnsString()
    {
        $this->picturelink = new Picturelink();
        $actual = $this->picturelink->makeLink('/img/mypicture');
        $this->assertInternalType('string', $actual);
    }

    public function testMakeLinkWithOneImage()
    {
        $url = "/img/test.jpg";
        $sizes = array(320);

        $expected = "<picture>".PHP_EOL;
        $expected .= '<!--[if IE 9]><video style="display: none;"><![endif]-->'.PHP_EOL;
        $expected .= '<source media="(max-width: 320px)" srcset="/img/test320.jpg" />'.PHP_EOL;
        $expected .= '<!--[if IE 9]></video><![endif]-->'.PHP_EOL;
        $expected .= '<img src="/img/test320.jpg" />'.PHP_EOL;
        $expected .= '</picture>'.PHP_EOL;

        $actual = $this->picturelink->makeLink($url, $sizes);
        $this->assertStringStartsWith($expected, $actual);
    }
    public function testMakeLinkWithMultipleImages()
    {
        $url = "/img/test.jpg";
        $sizes = array(320, 480, 768, 1024, 1224, 1824);
        $expected = "<picture>".PHP_EOL;
        $expected .= '<!--[if IE 9]><video style="display: none;"><![endif]-->'.PHP_EOL;
        $expected .= '<source media="(max-width: 320px)" srcset="/img/test320.jpg" />'.PHP_EOL;
        $expected .= '<source media="(max-width: 480px)" srcset="/img/test480.jpg" />'.PHP_EOL;
        $expected .= '<source media="(max-width: 768px)" srcset="/img/test768.jpg" />'.PHP_EOL;
        $expected .= '<source media="(max-width: 1024px)" srcset="/img/test1024.jpg" />'.PHP_EOL;
        $expected .= '<source media="(max-width: 1224px)" srcset="/img/test1224.jpg" />'.PHP_EOL;
        $expected .= '<source media="(min-width: 1224px)" srcset="/img/test1824.jpg" />'.PHP_EOL;

        $expected .= '<!--[if IE 9]></video><![endif]-->'.PHP_EOL;
        $expected .= '<img src="/img/test1824.jpg" />'.PHP_EOL;
        $expected .= '</picture>'.PHP_EOL;

        $actual = $this->picturelink->makeLink($url, $sizes);
        $this->assertStringStartsWith($expected, $actual);
    }
    public function testMakeLinkWithMultipleImagesRandomOrder()
    {
        $url = "/img/test.jpg";
        $sizes = array(1224, 768, 320, 1824, 480, 1024);
        $expected = "<picture>".PHP_EOL;
        $expected .= '<!--[if IE 9]><video style="display: none;"><![endif]-->'.PHP_EOL;
        $expected .= '<source media="(max-width: 320px)" srcset="/img/test320.jpg" />'.PHP_EOL;
        $expected .= '<source media="(max-width: 480px)" srcset="/img/test480.jpg" />'.PHP_EOL;
        $expected .= '<source media="(max-width: 768px)" srcset="/img/test768.jpg" />'.PHP_EOL;
        $expected .= '<source media="(max-width: 1024px)" srcset="/img/test1024.jpg" />'.PHP_EOL;
        $expected .= '<source media="(max-width: 1224px)" srcset="/img/test1224.jpg" />'.PHP_EOL;
        $expected .= '<source media="(min-width: 1224px)" srcset="/img/test1824.jpg" />'.PHP_EOL;

        $expected .= '<!--[if IE 9]></video><![endif]-->'.PHP_EOL;
        $expected .= '<img src="/img/test1824.jpg" />'.PHP_EOL;
        $expected .= '</picture>'.PHP_EOL;

        $actual = $this->picturelink->makeLink($url, $sizes);
        $this->assertStringStartsWith($expected, $actual);
    }

    public function testMakeLinkWithOneImageAnd2xResolution()
    {
        $url = "/img/test.jpg";
        $sizes = array(320);
        $option = array('resolution' => '2x');

        $expected = "<picture>".PHP_EOL;
        $expected .= '<!--[if IE 9]><video style="display: none;"><![endif]-->'.PHP_EOL;
        $expected .= '<source media="(max-width: 320px)" srcset="/img/test320.jpg, /img/test320@2x.jpg 2x" />'.PHP_EOL;
        $expected .= '<!--[if IE 9]></video><![endif]-->'.PHP_EOL;
        $expected .= '<img src="/img/test320.jpg" />'.PHP_EOL;
        $expected .= '</picture>'.PHP_EOL;

        $actual = $this->picturelink->makeLink($url, $sizes, $option);
        $this->assertStringStartsWith($expected, $actual);
    }

    public function testMakeLinkWithOneImageAndResolutionArray()
    {
        $url = "/img/test.jpg";
        $sizes = array(320);
        $option = array('resolution' => array('2x','3x'));

        $expected = "<picture>".PHP_EOL;
        $expected .= '<!--[if IE 9]><video style="display: none;"><![endif]-->'.PHP_EOL;
        $expected .= '<source media="(max-width: 320px)" srcset="/img/test320.jpg, /img/test320@2x.jpg 2x, /img/test320@3x.jpg 3x" />'.PHP_EOL;
        $expected .= '<!--[if IE 9]></video><![endif]-->'.PHP_EOL;
        $expected .= '<img src="/img/test320.jpg" />'.PHP_EOL;
        $expected .= '</picture>'.PHP_EOL;

        $actual = $this->picturelink->makeLink($url, $sizes, $option);
        $this->assertStringStartsWith($expected, $actual);
    }

    /**
     * @expectedException InvalidArgumentException
     */
    public function testMakeLinkWithNonNumbersAsSizes(){
        $url = '/img/test.jpg';
        $sizes = array('A');
        $this->picturelink->makeLink($url, $sizes);
    }

    public function testMakeLinkWithOneAttribute()
    {
        $url = "/img/test.jpg";
        $sizes = array(320);
        $attributes = array('id' => 'testId');

        $expected = '<picture id="testId" >'.PHP_EOL;
        $expected .= '<!--[if IE 9]><video style="display: none;"><![endif]-->'.PHP_EOL;
        $expected .= '<source media="(max-width: 320px)" srcset="/img/test320.jpg" />'.PHP_EOL;
        $expected .= '<!--[if IE 9]></video><![endif]-->'.PHP_EOL;
        $expected .= '<img src="/img/test320.jpg" />'.PHP_EOL;
        $expected .= '</picture>'.PHP_EOL;

        $actual = $this->picturelink->makeLink($url, $sizes, array(), $attributes);
        $this->assertStringStartsWith($expected, $actual);
    }

    public function testMakeLinkWithMultipleAttributes()
    {
        $url = "/img/test.jpg";
        $sizes = array(320);
        $attributes = array('id' => 'testId', 'class' => 'test-class');

        $expected = '<picture id="testId" class="test-class" >'.PHP_EOL;
        $expected .= '<!--[if IE 9]><video style="display: none;"><![endif]-->'.PHP_EOL;
        $expected .= '<source media="(max-width: 320px)" srcset="/img/test320.jpg" />'.PHP_EOL;
        $expected .= '<!--[if IE 9]></video><![endif]-->'.PHP_EOL;
        $expected .= '<img src="/img/test320.jpg" />'.PHP_EOL;
        $expected .= '</picture>'.PHP_EOL;

        $actual = $this->picturelink->makeLink($url, $sizes, array(), $attributes);
        $this->assertStringStartsWith($expected, $actual);
    }

    public function testMakeLinkWithAltTag()
    {
        $url = "/img/test.jpg";
        $sizes = array(320);
        $alt = "test image";

        $expected = '<picture>'.PHP_EOL;
        $expected .= '<!--[if IE 9]><video style="display: none;"><![endif]-->'.PHP_EOL;
        $expected .= '<source media="(max-width: 320px)" srcset="/img/test320.jpg" />'.PHP_EOL;
        $expected .= '<!--[if IE 9]></video><![endif]-->'.PHP_EOL;
        $expected .= '<img src="/img/test320.jpg" alt="test image" />'.PHP_EOL;
        $expected .= '</picture>'.PHP_EOL;

        $actual = $this->picturelink->makeLink($url, $sizes, array(), array(), $alt);
        $this->assertStringStartsWith($expected, $actual);
    }
}


gulpfile.js
var gulp = require('gulp');
var responsive = require('gulp-responsive');

gulp.task('img', function(){
    return gulp.src('/img/**/*')
        .pipe(
        responsive(
            {
                '**/*' : [
                    {
                        width: 320,
                        withoutEnlargement: false,
                        rename: {
                            //prefix: "50-x-",
                            suffix: "320"
                            //path.extname = ".md"
                        }
                    },
                    {
                        width: 320 * 2,
                        withoutEnlargement: false,
                        rename: {
                            suffix: "320@2x"
                        }
                    },
                    {
                        width: 320 * 3,
                        withoutEnlargement: false,
                        rename: {
                            suffix: "320@3x"
                        }
                    },
                    {
                        width: 480,
                        withoutEnlargement: false,
                        rename: {
                            suffix: "480"
                        }
                    },
                    {
                        width: 480 * 2,
                        withoutEnlargement: false,
                        rename: {
                            suffix: "480@2x"
                        }
                    },
                    {
                        width: 480 * 3,
                        withoutEnlargement: false,
                        rename: {
                            suffix: "480@3x"
                        }
                    },
                    {
                        width: 768,
                        withoutEnlargement: false,
                        rename: {
                            suffix: "768"
                        }
                    },
                    {
                        width: 768 * 2,
                        withoutEnlargement: false,
                        rename: {
                            suffix: "768@2x"
                        }
                    },
                    {
                        width: 768 * 3,
                        withoutEnlargement: false,
                        rename: {
                            suffix: "768@3x"
                        }
                    },
                    {
                        width: 1024,
                        withoutEnlargement: false,
                        rename: {
                            suffix: "1024"
                        }
                    },
                    {
                        width: 1024 * 2,
                        withoutEnlargement: false,
                        rename: {
                            suffix: "1024@2x"
                        }
                    },
                    {
                        width: 1024 * 3,
                        withoutEnlargement: false,
                        rename: {
                            suffix: "1024@3x"
                        }
                    },
                    {
                        width: 1224,
                        withoutEnlargement: false,
                        rename: {
                            suffix: "1224"
                        }
                    },
                    {
                        width: 1224 * 2,
                        withoutEnlargement: false,
                        rename: {
                            suffix: "1224@2"
                        }
                    },
                    {
                        width: 1224 * 3,
                        withoutEnlargement: false,
                        rename: {
                            suffix: "1224@3"
                        }
                    },
                    {
                        width: 1824,
                        withoutEnlargement: false,
                        rename: {
                            suffix: "1824"
                        }
                    },
                    {
                        width: 1824 * 2,
                        withoutEnlargement: false,
                        rename: {
                            suffix: "1824@2x"
                        }
                    },
                    {
                        width: 1824 * 3,
                        withoutEnlargement: false,
                        rename: {
                            suffix: "1824@3x"
                        }
                    },
                ]
            })
    )
        .pipe(gulp.dest('/img'));
});

gulp.task('watch:img', function(){
    gulp.watch('/src/img/*',['img']);
});



src/Picturelink.php
<?php
/**
 * Created by PhpStorm.
 * User: Dag
 * Date: 08/06/2015
 * Time: 12:39
 */

namespace Divarsoy\Picturelink;

class Picturelink {

    /**
     * Creates a Picturelink with srcsets
     * @param string $url
     * @param array $sizes
     * @param array $options
     * @param string $alt
     * @param array $attributes
     * @return string
     */
    public function makeLink($url, array $sizes = array(), array $options = array(), $attributes = array(), $alt = null){
        $returnString = '';
        $returnString .= $this->makePictureStartTag($attributes);
        $returnString .= $this->makeIEfixStart();
        $returnString .= $this->makeSource($url, $sizes, $options);
        $returnString .= $this->makeIEfixEnd();
        $returnString .= $this->makeFallback($url, $sizes, $alt);
        $returnString .= $this->makePictureEndTag();
        return $returnString;
    }

    /**
     * Creates a <Picture> tag with attributes
     *
     * @param array $attributes
     * @return string
     */
    protected function makePictureStartTag($attributes = array())
    {
        return "<picture".$this->attributes($attributes). ">".PHP_EOL;
    }

    /**
     * Creates a </picture> tag
     *
     * @return string
     */
    protected function makePictureEndTag()
    {
        return "</picture>".PHP_EOL;
    }

    /**
     * Creates a workaround for IE9
     *
     * @return string
     */
    protected function makeIEfixStart()
    {
        return '<!--[if IE 9]><video style="display: none;"><![endif]-->'.PHP_EOL;
    }

    /**
     * Creates an endtag for the IE9 workaround
     *
     * @return string
     */
    protected function makeIEfixEnd()
    {
        return '<!--[if IE 9]></video><![endif]-->'.PHP_EOL;
    }

    /**
     * Build an HTML attribute string from an array.
     *
     * @param  array  $attributes
     * @return string
     */
    protected function attributes($attributes)
    {
        if (!empty($attributes)) {
            $returnString = ' ';

            foreach ($attributes as $key => $value) {
                $returnString .= $key . '="' . $value . '" ';
            }
            return $returnString;
        }
    }

    /**
     * Gets the extension part from an url to an image
     *
     * @param string $url
     * @return string
     */
    protected function getExtension($url){
        return pathinfo($url, PATHINFO_EXTENSION);
    }

    /**
     * Gets the url without the extension
     *
     * @param string $url
     * @return string
     */
    protected function removeExtension($url){
        $path = pathinfo($url);
        return $path['dirname'].'/'.$path['filename'];
    }

    /**
     * Makes the Srcsets from the $sizes array.
     *
     * @param string $url
     * @param array $sizes
     * @param array $option
     * @return string
     */
    protected function makeSource($url, array $sizes, array $option){
        $returnString = '';
        sort($sizes, SORT_NUMERIC);

        if (count($sizes) == 1){
            return $this->makeMaxWidthSource($url, $sizes[0], $option);
        }
        for($i=0; $i < count($sizes) - 1; $i++ ) {
            $returnString .= $this->makeMaxWidthSource($url, $sizes[$i], $option);
        }
        if (count($sizes) > 2) {
            $lastSize = $sizes[count($sizes) - 1];
            $nextToLastSize = $sizes[count($sizes) - 2];

            $returnString .= $this->makeMinWidthSource($url, $option, $lastSize, $nextToLastSize);
        }
        return $returnString;
    }

    /**
     * @param $url
     * @param int $size
     * @param array $option
     * @return string
     */
    protected function makeMaxWidthSource($url, $size, array $option)
    {
        $this->checkIfNumber($size);
        $returnString = '';
        $returnString .= '<source media="(max-width: ' . $size . 'px)" ';
        $returnString .= 'srcset="' . $this->removeExtension($url) . $size . '.' . $this->getExtension($url);
        $returnString .= $this->makeHighResolutionLinks($url, $size, $option);
        $returnString .= '" />' . PHP_EOL;
        return $returnString;
    }
    /**
     * @param $url
     * @param array $option
     * @param int $size
     * @param int $minSize
     * @return string
     */
    protected function makeMinWidthSource($url, array $option, $size, $minSize)
    {
        $returnString = '';
        $this->checkIfNumber($size);
        $returnString .= '<source media="(min-width: ' . $minSize . 'px)" ';
        $returnString .= 'srcset="' . $this->removeExtension($url) . $size . '.' . $this->getExtension($url);
        $returnString .= $this->makeHighResolutionLinks($url, $size, $option);
        $returnString .= '" />';

        return $returnString . PHP_EOL;
    }

    /**
     * Checks that the size specified is a number
     * @param $number
     */
    protected function checkIfNumber($number){
        if(! is_numeric($number)){
            throw new \InvalidArgumentException(
                'Size array must only consist of numbers'
            );
        }
    }

    /**
     * Makes @2x srcsets if set in options
     *
     * @param string $url
     * @param int $size
     * @param array $options
     * @return string
     */
    protected function makeHighResolutionLinks($url, $size, array $options){
        $returnString = '';
        if (isset($options['resolution'])){
            if( is_array($options['resolution'])) {
                foreach ($options['resolution'] as $resolution) {
                    $returnString .= ', ' . $this->removeExtension($url) . $size . '@' . $resolution . '.' . $this->getExtension($url) . ' ' . $resolution;
                }
                return $returnString;
            }
            else {
                return ', ' . $this->removeExtension($url) . $size . '@' . $options['resolution'] . '.' . $this->getExtension($url) . ' ' . $options['resolution'];
            }
        }
        else {
            return '';
        }
    }

    /**
     * Makes image fallback
     *
     * @param string $url
     * @param array $sizes
     * @param string $alt
     * @return string
     */
    protected function makeFallback($url, array $sizes, $alt = null){
        sort($sizes);
        if(!empty($sizes)) {
            $returnString = '<img src="' . $this->removeExtension($url) . $sizes[count($sizes) - 1] . '.' . $this->getExtension($url) . '" ';

            if (isset($alt)) {
                $returnString .= 'alt="' . $alt . '" ';
            }
            $returnString .= '/>';

            return $returnString . PHP_EOL;
        }
    }
}